### PR TITLE
Add some ansible modules and plugins to  manage HUAWEI Router.

### PR DESCRIPTION
**Pull request by meijun3**
…weirouters.

##### SUMMARY

Using there modules,Ansible can manage VRP-based ATN,NE,ME and CX series routers in Huawei data centers.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

- Add plugins:
plugins\action\ne.py
plugins\action\ne_config.py
plugins\cliconf\ne.py
plugins\netconf\ne.py
plugins\terminal\ne.py
- Add modules:
Modules in modules\network\ne ,include acl,bgp,interface(ifm),lldp,isis,mpls,qos etc.
- Add module_utils:
module_utils\network\ne\ne.py
- Modified files:
plugins\action\net_base.py
plugins\connection\netconf.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by meijun3** [comment]
Shippable — Run 100148 status is FAILED.  
--- Should i fix this failed ?

File: lib/ansible/plugins/cliconf/ne.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    @enable_mode
+    def get_config(self, source='running'):
+        if source != 'running':
+            return self.invalid_params("fetching configuration from %s is not supported" % source)
+        cmd = 'display current-configuration'
+        return self.send_command(cmd)
+
+    @enable_mode
+    def edit_config(self, command):
+        for cmd in chain([b'configure terminal'], to_list(command), [b'end']):
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
        for cmd in chain(['configure terminal'], to_list(command), ['end']):
```

File: lib/ansible/plugins/cliconf/ne.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+from ansible.plugins.cliconf import CliconfBase, enable_mode
+
+
+class Cliconf(CliconfBase):
+
+    def get_device_info(self):
+        device_info = dict()
+
+        device_info['network_os'] = 'ne'
+        reply = self.get(b'display version')
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
        reply = self.get('display version')
```

File: lib/ansible/plugins/cliconf/ne.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+#
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import re
+import json
+
+from itertools import chain
+
+from ansible.module_utils._text import to_bytes, to_text
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
from ansible.module_utils._text import to_text
```

**Review by Qalthos — CHANGES_REQUESTED** [review]
You shouldn't need to have nearly as many raw bytes as you're using

File: lib/ansible/plugins/cliconf/ne.py (hunk 1 above)
Comment (meijun3) [reply lib/ansible/plugins/cliconf/ne.py]: we have updated the file,check again please,thinks.
```suggestion
        for cmd in chain([b'configure terminal'], to_list(command), [b'end']):
```

**Comment by meijun3** [comment]
@Qalthos  hello，I have updated the file, please check again,thanks.

**Comment by meijun3** [comment]
@Qalthos @ganeshrn @justjais Hello all,if you finish to review this PR,and it is passed,could you please merge this PR? Thanks.

File: lib/ansible/plugins/cliconf/ne.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    @enable_mode
+    def edit_config(self, command):
+        for cmd in chain(['configure terminal'], to_list(command), ['end']):
+            self.send_command(cmd)
+
+    def get(self, command, prompt=None, answer=None, sendonly=False):
+        return self.send_command(command, prompt=prompt, answer=answer, sendonly=sendonly)
+
+    def get_capabilities(self):
+        result = {}
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
        result = super(Cliconf, self).get_capabilities()
```

File: lib/ansible/plugins/cliconf/ne.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        result['rpc'] = self.get_base_rpc()
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
```

File: lib/ansible/plugins/cliconf/ne.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        result['network_api'] = 'cliconf'
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
```

File: lib/ansible/plugins/cliconf/ne.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        result['device_info'] = self.get_device_info()
```
Comment (Qalthos) [inline lib/ansible/plugins/cliconf/ne.py]:
```suggestion
```

File: lib/ansible/plugins/terminal/ne.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import re
+
+from ansible.plugins.terminal import TerminalBase
+from ansible.errors import AnsibleConnectionFailure
+
+
+class TerminalModule(TerminalBase):
+
+    terminal_stdout_re = [
+        re.compile(r'[\r\n]?<.+>(?:\s*)$'),
```
Comment (Qalthos) [inline lib/ansible/plugins/terminal/ne.py]:
```suggestion
        re.compile(br'[\r\n]?<.+>(?:\s*)$'),
```

And so on for all the entries of both `terminal_stdout_re` and `terminal_stderr_re`.

File: lib/ansible/plugins/terminal/ne.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        re.compile(r"'[^']' +returned error code: ?\d+"),
+        re.compile(r"syntax error"),
+        re.compile(r"unknown command"),
+        re.compile(r"Error\[\d+\]: ", re.I),
+        re.compile(r"Error:", re.I)
+    ]
+
+    def on_open_shell(self):
+        try:
+            self._exec_cli_command('screen-length 0 temporary')
```
Comment (Qalthos) [inline lib/ansible/plugins/terminal/ne.py]:
```suggestion
            self._exec_cli_command(b'screen-length 0 temporary')
```

**Review by Qalthos — CHANGES_REQUESTED** [review]
Some changes have gone in to cliconf, so there's some things no longer necessary there.

Additionally, terminal runs lower than cliconf, so those strings _do_ need to be specified as byte strings

**Comment by Qalthos** [comment]
Thank you very much for your interest in Ansible. This plugin will not be maintained in this repository and must be migrated to https://github.com/ansible-collections/community.network

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Pull request closed without merging**